### DISCUSSION TITLE
error: stray ‘\345’ in program

### DISCUSSION THREAD

**Discussion by zyjdmmm** [xmake-io/xmake#3218] [Q&A]
### Xmake 版本

2.7.4

### 操作系统版本和架构

centos 7.8

### 描述问题

按照如下步骤能稳定复现c程序的注释报错的bug：
1 、执行xmake create hello
2 、cd hello
3、xmake
4、打开xmake.lua修改为
```
add_rules("mode.debug", "mode.release")
set_languages("c89", "c++11")

target("hello")
    set_kind("binary")
    add_files("src/*.cpp")
    add_files("src/*.c")
```

5、把main.cpp修改为main.c
6、随意复制一个c程序，只要是带注释即可
```
#include<unistd.h> //和fd有关[read() write() close()] lseek fork() execl() dup2() getpid()
#include<fcntl.h>//open()
#include<stdio.h>//perror()
#include<stdlib.h> //exit()
#include<dirent.h> //目录项操作：DIR opendir closedir readdir
#include<sys/wait.h> //wait()

#include<string.h>

int main(int argc,char *argv[])
{
    int ret;
    int fd[2];
    pid_t pid;
    char str[]="hello pipe\n";
    char buf[1024]={0};
    int status,w_pid;

    //创建管道
    ret=pipe(fd);
    if(-1 == ret){
        perror("pipe error");
    }

    pid=fork();
    if(pid>0){//父进程
        close(fd[0]);//关闭读端
        dup2(fd[1],STDOUT_FILENO);//屏幕输出重定向到管道 写入
        execlp("ls","ls",NULL);

        perror("exclp error");
        //因为使用exec函数族执行其他程序成功就不会回来了，所以下面就不用写了，只能指望隐式回收
        // close(fd[1]);
        // //阻塞等待子进程死亡防止僵尸进程
        // w_pid=wait(&status);
        // if(-1==w_pid){
        //     perror("wait error");
        //     exit(1);
        // }
    }else if(0==pid){//子进程
        close(fd[1]);//关闭写端
        dup2(fd[0],STDIN_FILENO);//从屏幕读取重定向到管道 读取
        execlp("wc","wc","-l",NULL);
        perror("exclp error");

        //执行成功不再执行以下的
        close(fd[0]);
    }

    return 0;
}
```

7、发现只要带注释的行都报错，有游离的"/"
8、发现是c标准的问题,改为c90后编译bug消失
9、但某些情况c89又能正常编译，忘了怎么触发的

### 期待的结果

报错如下
```
[xmk@VM-4-3-centos hello_xmake]$ xmake
[ 25%]: cache compiling.release src/main.c
error: src/main.c:19:5: error: expected expression before ‘/’ token
     //创建管道
     ^
src/main.c:19:5: error: stray ‘\345’ in program
src/main.c:19:5: error: stray ‘\210’ in program
src/main.c:19:5: error: stray ‘\233’ in program
src/main.c:19:5: error: stray ‘\345’ in program
src/main.c:19:5: error: stray ‘\273’ in program
src/main.c:19:5: error: stray ‘\272’ in program
src/main.c:19:5: error: stray ‘\347’ in program
src/main.c:19:5: error: stray ‘\256’ in program
src/main.c:19:5: error: stray ‘\241’ in program
src/main.c:19:5: error: stray ‘\351’ in program
src/main.c:19:5: error: stray ‘\201’ in program
src/main.c:19:5: error: stray ‘\223’ in program
src/main.c:26:15: error: expected expression before ‘/’ token
     if(pid>0){//父进程
  > in src/main.c
warning: cannot match target(hello).add_files("src/*.cpp") at ./xmake.lua:6
```

### 工程配置

```
add_rules("mode.debug", "mode.release")
set_languages("c89", "c++11")

target("hello")
    set_kind("binary")
    add_files("src/*.cpp")
    add_files("src/*.c")
```

**Comment by waruqi** [comment]
> 7、发现只要带注释的行都报错，有游离的"/"
8、发现是c标准的问题,改为c90后编译bug消失

这不是 bug 把，编译器都报错了，说明 c89 压根不支持 `//`，跟 bug 没啥关系，而且即使是 bug ，也是编译器的问题，跟 xmake 更没啥关系

src/main.c:20:5: error: C++ style comments are not allowed in ISO C90

看错误提示，`//` 是 c++ 风格注释, c89 不支持。

你改成 c99 就行了。
 
9、但某些情况c89又能正常编译，忘了怎么触发的

没遇到过。

**Reply by zyjdmmm** [reply]
好的谢谢，才疏学浅了

**Comment by zyjdmmm** [comment]
我发现c89是支持//的。

使用以下方法稳定触发error: stray ‘\345’ in program：
1、使用如下配置和上述main.c
```
add_rules("mode.debug", "mode.release")
set_languages("c89", "c++11")

target("hello")
    set_kind("binary")
    add_files("src/*.cpp")
    add_files("src/*.c")
    add_syslinks("pthread")
```
2、先点vscode build编译一个文件hello（此时能编译）
3、修改目标名字target("hello123")再vscode build编译第2个（此时还能编译）
4、修改目标名字target("hello12345")再vscode build编译第3个(此时报错src/main.c:28:24: error: stray)

反正按上述流程能触发报错，我也觉得很奇怪。。。
@waruqi

**Reply by waruqi** [reply]
我这编译器都明显提示不支持了 
src/main.c:20:5: error: C++ style comments are not allowed in ISO C90

**Reply by zyjdmmm** [reply]
可是我这里是支持的的，没遇到你那种报错
```
[root@VM-4-3-centos ~]# gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Reply by waruqi** [reply]
你禁用下 cache 试下，xmake f --ccache=n 

如果还是有问题，那跟 xmake 没啥关系，即使 gcc 支持，报错，也是编译器的问题，你得到 gcc 那边提 issues ，而不是这里

也可以命令行复制出来，直接运行测试

**Reply by zyjdmmm** [reply]
那我还是用c90把。

那么在vscode中
```
set_languages("c90", "c++11")
```
能不能增加一个类似于/hello_x/.vscode/c_cpp_properties.json文件夹中自动弹出c/c++标准选项的功能？
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "compilerPath": "/usr/lib64/ccache/gcc",
            "cStandard": "gnu89",
            "cppStandard": "c++14",
            "intelliSenseMode": "linux-gcc-x64"
        }
    ],
    "version": 4
}
```
 "cStandard": "gnu89",键入有提示感觉很人性化，而xmake中我不太确定c标准支持哪些：gnu89还有gun90？甚至写c70也没有报错，不知道程序编译究竟用了哪个标准。如果不行的话至少给个warning呐

谢谢。

**Reply by waruqi** [reply]
https://github.com/xmake-io/xmake/blob/de6ac8dde75789da3164494cda26d0c402551d99/xmake/modules/core/tools/gcc.lua#L184

**Reply by zyjdmmm** [reply]
好的，希望未来更好用些

**Comment by zyjdmmm** [comment]
1、这个问题我后来弄懂了，这个标准和gcc的-std=接受的参数有关

#标准c系列：可填写c89 c90 c99 …（特别注意c89和c90不支持c++风格”//”注释，且这一系列标准比gnu系列少了很多函数扩展,而且这个注释报错在mingw上和gcc报错似乎不一样但总的来说就是不支持）
#linux下编程gnu系列：可填写为gnu89 gnu90 gnu99 …

2、-std=gnu90或-std=c90是可以的但我看gcc.lua里没有写？不太懂lua

3、gcc参数-std= 传入错误标准是直接会报错的，希望作者未来能给出错误提示

**Reply by waruqi** [reply]
> #标准c系列：可填写c89 c90 c99 …（特别注意c89和c90不支持c++风格”//”注释，且这一系列标准比gnu系列少了很多函数扩展,而且这个注释报错在mingw上和gcc报错似乎不一样但总的来说就是不支持）

我之前就说了不支持，错误提示很明显

> 2、-std=gnu90或-std=c90是可以的但我看gcc.lua里没有写？不太懂lua

dev 加上了

3、gcc参数-std= 传入错误标准是直接会报错的，希望作者未来能给出错误提示

> src/main.c:20:5: error: C++ style comments are not allowed in ISO C90

我这里编译器不是报了么，还要提示什么？没法复现你的问题

**Reply by zyjdmmm** [reply]
关于第3点，例如传入 拼写错误的gun 和 错误的年份91 都不会报错，现在xmake采用了什么标准就令人很困惑。
`set_languages("gun91", "c++11")`

**Reply by waruqi** [reply]
https://github.com/xmake-io/xmake/issues/3320

后面会搞个专门的 `xmake check` 诊断工具去检测各类写法错误

**Reply by zyjdmmm** [reply]
好的

**Reply by waruqi** [reply]
加上了，不过还没 merge 下个版本支持，具体进度见

https://github.com/xmake-io/xmake/issues/3326

**Reply by zyjdmmm** [reply]
ok!
